src/java/ognl/security/MethodBodyExecutionSandbox.java

```diff
@@ -0,0 +1,87 @@
+package ognl.security;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+/**
+ * Guarantee a singleton shared thread-safe security manager and sandbox for user's methods body execution.
+ *
+ * Add the `-Dognl.security.manager` to JVM options to enable.
+ *
+ * <p> Note: Due to potential performance and concurrency issues, try this only if you afraid your app can have an
+ * unknown "expression injection" flaw or you afraid you cannot prevent those in your app's internal sandbox
+ * comprehensively e.g. you cannot discover and maintain all attack vectors over time because of many dependencies
+ * and also their change over time.</p>
+ *
+ * <p> This tries to provide an option to you to enable a security manager that disables any sensitive action e.g.
+ * exec and exit even if attacker had a successful "expression injection" in any unknown way into your app. However,
+ * also honors previous security manager and policies if any set, as parent, and rolls back to them after method
+ * execution finished.</p>
+ *
+ * @author Yasser Zamani
+ * @since 3.1.23
+ */
+public class MethodBodyExecutionSandbox {
+    private static int residentsCount = 0;
+
+    private static SecurityManager parentSecurityManager;
+
+    public static Object executeMethodBody(Object target, Method method, Object[] argsArray) throws InvocationTargetException,
+            IllegalAccessException {
+        
+        if (!enter()) {
+            // isn't enabled or cannot installed due to current security manager policy
+            return method.invoke(target, argsArray);
+        }
+
+        try {
+            return method.invoke(target, argsArray);
+        } finally {
+            leave();
+        }
+    }
+
+    private static boolean enter() {
```

Comment:
Suggestion:  Move `getProperty()` check outside of synchronization block for better performance:
```java
                if (System.getProperty("ognl.security.manager") == null) {
                    // isn't enabled so simply just invoke the method outside sandbox
                    return false;
                }
                synchronized (MethodBodyExecutionSandbox.class) {
                    if (installSandboxIntoJVM()) {
                // Rest of code
```
If the `sandboxEnabled` flag idea is used, replace the `if (getProperty())` check with `if (!sandboxEnabled)`.

Reply:
It must be inside `if (residentsCount == 0) {` elsewhere it throws exception - already our own installed security manager doesn't allow reading any property :) -- `residentsCount == 0` means there is no already our own installed security manager.

Reply:
Hello @yasserzamani .

If I'm reading the code correctly, if `System.getProperty("ognl.security.manager")` has to be inside `synchronized (MethodBodyExecutionSandbox.class) ` it's _going to force a thread-synchronization before every OGNL method invocation_ ... _even if the library user doesn't want the sandbox enabled_ (system property false).  :(

From what I can see that creates an issue of overhead/thread synchronization that the library user has to pay even if they have no intention of ever using the sandbox.  If it's turned off then the implementation shouldn't have any extra synchronization cost over previous ONGNL, should it ?

There is also the issue of unexpected side-effects from the more-restrictive sandbox security manager running while any OGNL expression invocations are taking place ... which could break non-OGNL code running in the same JVM while it's in effect (in hard-to-track ways) ... unless I'm missing something.   :(

It seems like the _sandbox may be too far-reaching for its intended purpose_, but maybe I'm overlooking or misunderstanding.  :confused:

Maybe a sandboxed OGNL should be a parallel fork (so it can be swapped-in/swapped-out at the .jar level) ?  Or maybe a duplicate-but-separate `OgnlRuntimeSandboxed` that the library user decides at compile or runtime to use (instead of the standard `OgnlRuntime`) ?  :confused:

Reply:
👍 Great catch! I'll move it outside synchronized block, thanks! - while it has a surrounding try-catch now, it's possible now to move it out :)

I think it doesn't break non-ognl code because of `isAccessDenied` check but you're right, it's currently JVM wide and currently I think it has to be because JDK's one is JVM wide - I understand your concern for example user has two web apps in same JVM and want sandbox for one and don't want for another. To address this I guess we should include `ProtectionDomain` concept in `isAccessDenied` method.

Yes I understand it's complexity and huge side-effects but I wasn't able a better complete robust solution for "expression injection". However, I feel we will handle all at end via improving `isAccessDenied` method logic - this PR gets better every day :) And **as far as I know OGNL, I generally see it rational for OGNL to disable sensitive actions at all because in case of an unknown "expression injection" flaw in user app, the attacker has a powerful language and compiler in hand to do anything! @ ognl_community  right?**

No I think all of these are finally handle-able via a good `isAccessDenied` method logic :)